docs/design/ConstructiveSolidGeometry.MD

```diff
@@ -0,0 +1,136 @@
+
+`Intersection` - visually the part of A which intersects (lies within) B and vice versa. The intersected shapes is contained in *both* A and B. If the two basic shapes do not intersect (do not touch each other), the result is an empty shape. In the example above the red and blue shapes are intersected.
+
+`Subtraction` - visually subtracts (carves out) B from A. If the shapes do not touch, there is nothing to subtract so the result is the same as A. If B fully contains A, subtracting it will remove A entirely leaving just an empty shape. In the example above, the green part is subtracted from the red-blue intersection.
+
+## The List
+
+In order to traverse the tree easily we flatten it to a list and iterate over the list linearly.
+
+1. Make sure we are working with a binary tree. If a node unifies or intersects three or more shapes, the node should be split to a binary union or intersection of the first shape and a new operator node (union or intersection) of the rest of the shapes. Repeat until all operators are binary.
+
+2. Traverse the tree depth-first. Add each node on *leaving* the node. In the example above:
+ - Enter the root (subtraction) node.
+ - Enter the left (intersection) node.
+ - Enter the red cube. It's a leaf node so we leave it immediately and add it to the list (1. Red cube)
+ - Enter the blue sphere. It's a leaf node so we leave it immediately and add it to the list (2. Blue sphere)
+ - Leave the intersection node and add it to the list (3. Intersect)
+ - Enter the right (union) node.
+ - Enter the left green cylinder. It's a leaf node so we leave it immediately and add it to the list (4. Green cylinder)
+ - Enter the second right (union) node.
+ - Enter the left green cylinder. It's a leaf node so we leave it immediately and add it to the list (5. Green cylinder)
+ - Enter the right green cylinder. It's a leaf node so we leave it immediately and add it to the list (6. Green cylinder)
+ - Leave the second right (union) node and add it to the list (7. Union)
+ - Leave the first right (union) node and add it to the list (8. Union)
+ - Leave the root (subtraction) node and add it to the list (9. Subtraction)
+
+The list now looks like this:
+```
+1. Red cube
+2. Blue sphere
+3. Intersect
+4. Green cylinder
+5. Green cylinder
+6. Green cylinder
+7. Union
+8. Union
+9. Subtraction
+```
+If we treat it as a Reverse Polish Notation we can iterate it using a simple stack.
+
+```
+1. -> [1. RC]
+2. -> [1. RC, 2. BS]
+3. -> [1. RC * BS]
+4. -> [1. RC * BS, 2. GC]
+5. -> [1. RC * BS, 2. GC, 3. GC]
+6. -> [1. RC * BS, 2. GC, 3. GC, 4. GC]
+7. -> [1. RC * BS, 2. GC, 3. GC + GC]
+8. -> [1. RC * BS, 2. GC + GC + GC]
+9. -> [1. RC * BS - (GC + GC + GC)]
+```
+
+## The Shader
```

Comment:
Explained very well till the last section (was able to follow quite easily).
From this section onwards it gets a bit dense for anyone who's never written any graphics code or doesn't have experience with 3D geometry math. But this is understandable since I assume the primary audience for this writeup is graphics programmers.

Reply:
Thanks! Let me know if anything from the last section is unclear though, I can go into more detail. But if the difficulty feels increasing with the paragraphs I think it's good; the doc is meant as a short explanation for the engineers, including people on our team working with the library.

Reply:
I agree, the doc works as intended and for the audience that it's aimed at.
Looks good!